-1 
Tested at: 20ea22ad338ac1527c013fcb9bffaecf866364aa
 I found an error when building: 
 
 <pre>>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2016-01-13-2300/src/RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc 
>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2016-01-13-2300/src/RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc 
>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2016-01-13-2300/src/RecoParticleFlow/PFProducer/src/PFEGammaHeavyObjectCache.cc 
>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2016-01-13-2300/src/RecoParticleFlow/PFProducer/src/PFElectronAlgo.cc 
/tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2016-01-13-2300/src/RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc: In member function 'void PFEGammaAlgo::mergeROsByAnyLink(std::list<PFEGammaAlgo::ProtoEGObject>&)':
/tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2016-01-13-2300/src/RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc:1457:56: error: 'LOGWARNING' was not declared in this scope
            LOGWARNING("PFEGammaAlgo::mergeROsByAnyLink") 
                                                        ^
>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2016-01-13-2300/src/RecoParticleFlow/PFProducer/src/PFMuonAlgo.cc 
>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2016-01-13-2300/src/RecoParticleFlow/PFProducer/src/PFPhotonAlgo.cc 
>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2016-01-13-2300/src/RecoParticleFlow/PFProducer/src/PhotonSelectorAlgo.cc 
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-12939/10498/summary.html

--- a/RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc
+++ b/RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc
@@ -1449,6 +1449,8 @@ initializeProtoCands(std::list<PFEGammaAlgo::ProtoEGObject>& egobjs) {
          // -- check that the front is still mergeable!
          if( !mergeTest(*roToMerge) ) {
             // back up one step and repartition(since we will increment later)
+           LOGWARN("PFEGammaAlgo::mergeROsByAnyLink") 
+             << "Merge-i-ness of ROs changed while doing the merge!";
            nomerge = std::partition(roToMerge--,ROs.end(),mergeTest);
            continue;
          }

--- a/RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc
+++ b/RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc
@@ -1447,13 +1447,15 @@ initializeProtoCands(std::list<PFEGammaAlgo::ProtoEGObject>& egobjs) {
        for( auto roToMerge = mergestart; roToMerge != nomerge; ++roToMerge) {
          //bugfix! L.Gray 14 Jan 2016 
          // -- check that the front is still mergeable!
-         if( !mergeTest(*roToMerge) ) {
-            // back up one step and repartition(since we will increment later)
-           LOGWARN("PFEGammaAlgo::mergeROsByAnyLink") 
-             << "Merge-i-ness of ROs changed while doing the merge!";
-           nomerge = std::partition(roToMerge--,ROs.end(),mergeTest);
-           continue;
-         }
+         if( thefront.ecalclusters.size() && roToMerge->ecalclusters.size() ) {
+           if( thefront.ecalclusters.front().first->clusterRef()->layer() !=   
+               roToMerge->ecalclusters.front().first->clusterRef()->layer() ) {
+             LOGWARN("PFEGammaAlgo::mergeROsByAnyLink") 
+               << "Tried to merge EB and EE clusters! Skipping!";
+             ROs.push_back(*roToMerge);
+             continue;
+           }
+         }         
          //end bugfix
 	 thefront.ecalclusters.insert(thefront.ecalclusters.end(),
 				      roToMerge->ecalclusters.begin(),

